internal/jimm/jujuauth/jwtgenerator.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2024 Canonical.
+
+// Package jwtgenerator generates JWT tokens to
+// authenticate and authorize messages to Juju controllers.
+// This package is more specialised than a generic
+// JWT token generator as it crafts Juju specific
+// permissions that are added as claims to the JWT
+// and therefore exists in JIMM's business logic layer.
+package jujuauth
+
+import (
+	"context"
+	"sync"
+
+	"github.com/juju/names/v5"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/jimmjwx"
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+// generatorDatabase specifies the database interface used by the
+// JWT generator.
+type generatorDatabase interface {
+	GetController(ctx context.Context, controller *dbmodel.Controller) error
+}
+
+// generatorAccessChecker specifies the access checker used by the JWT
+// generator to obtain user's access rights to various entities.
+type generatorAccessChecker interface {
+	GetUserModelAccess(context.Context, *openfga.User, names.ModelTag) (string, error)
+	GetUserControllerAccess(context.Context, *openfga.User, names.ControllerTag) (string, error)
+	GetUserCloudAccess(context.Context, *openfga.User, names.CloudTag) (string, error)
+	CheckPermission(context.Context, *openfga.User, map[string]string, map[string]interface{}) (map[string]string, error)
+}
+
+// jwtService specifies the service JWT generator uses to generate JWTs.
+type jwtService interface {
+	NewJWT(context.Context, jimmjwx.JWTParams) ([]byte, error)
+}
+
+// TokenGenerator provides the necessary state and methods to authorize a user and generate JWT tokens.
+type TokenGenerator struct {
+	database      generatorDatabase
+	accessChecker generatorAccessChecker
+	jwtService    jwtService
+
+	mu             sync.Mutex
+	accessMapCache map[string]string
+	mt             names.ModelTag
+	ct             names.ControllerTag
+	user           *openfga.User
+	callCount      int
+}
+
+// New returns a new JWTGenerator.
+func New(database generatorDatabase, accessChecker generatorAccessChecker, jwtService jwtService) TokenGenerator {
```

Comment:
i still think argument types of exported methods need to be exported

Reply:
I'm trying to think of a reason why and I think I see a good concrete reason. An internal type can be freely changed without impacting the users of a package. In this this case if we were to change the definition of one of these interfaces we would break users of the package so it should be exported to indicate that this is used by external users.